File: LinkedList/typescript/palindrome_linked_list.ts
/**
 * Definition for singly-linked list.
*/
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function isPalindrome(head: ListNode | null): boolean {
    let slow = head
    let fast = head
    while (fast!=null && fast.next!=null && slow!=null){
        slow = slow?.next
        fast = fast.next.next
    }
    let prev:ListNode|null = null
    while(slow!=null){
        const next_node = slow.next
        slow.next = prev
        prev = slow
        slow = next_node
    }
    let left:ListNode|null = head
    let right:ListNode|null = prev
    while(left!=null && right!=null){
        if(left.val!=right.val){
            return false
        }
        left = left.next
        right = right.next
    }
    return true
};

File: LinkedList/python/palindrome_linked_list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution(object):
   
    def isPalindrome(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: bool
        """
        slow = fast = head
        while fast!=None and fast.next!=None:
            slow = slow.next
            fast = fast.next.next
        prev = None
        while slow!=None:
            next_node = slow.next
            slow.next = prev
            prev = slow
            slow = next_node
        left = head
        right = prev
        while left!=None and right!=None:
            if left.val!=right.val:
                return False
            left = left.next
            right = right.next
        return True        

        